src/panels/CharmsPanel/CharmsPanel.tsx

```diff
@@ -47,16 +47,32 @@ export default function CharmsPanel({
       loading={isLoading}
     >
       <form onSubmit={handleSubmit}>
-        {charms.map((charm) => (
-          <div key={charm.url} className="p-form__group">
-            <RadioInput
-              id={charm.url}
-              label={`${charm.meta?.name} (rev: ${charm.revision})`}
-              checked={selectedCharm === charm.url}
-              onChange={() => setSelectedCharm(charm.url)}
-            />
-          </div>
-        ))}
+        {charms.map((charm) => {
+          const hasActionData =
+            !!charm?.actions?.specs &&
+            !!Object.keys(charm.actions.specs).length;
+          return (
+            <div key={charm.url} className="p-form__group">
+              {hasActionData ? (
+                <RadioInput
+                  id={charm.url}
+                  label={`${charm.meta?.name} (rev: ${charm.revision})`}
+                  checked={selectedCharm === charm.url}
+                  onChange={() => setSelectedCharm(charm.url)}
+                />
+              ) : (
+                <Tooltip message={Label.NO_ACTIONS} position="left">
+                  <RadioInput
+                    id={charm.url}
+                    label={`${charm.meta?.name} (rev: ${charm.revision})`}
+                    checked={selectedCharm === charm.url}
+                    disabled={true}
+                  />
+                </Tooltip>
+              )}
```

Comment:
I think we can combine these. When there is no `message` passed to the tooltip, it will just render the contents, so I think it would be fine to always include it. Maybe something like this:

```suggestion
              <Tooltip
                message={hasActionData ? null : Label.NO_ACTIONS}
                position="left"
              >
                <RadioInput
                  id={charm.url}
                  label={`${charm.meta?.name} (rev: ${charm.revision})`}
                  checked={selectedCharm === charm.url}
                  onChange={() => setSelectedCharm(charm.url)}
                  disabled={!hasActionData}
                />
              </Tooltip>
```

Or the alternative would be to assign the input to a variable:

```tsx
const input = <RadioInput
                  id={charm.url}
                  label={`${charm.meta?.name} (rev: ${charm.revision})`}
                  checked={selectedCharm === charm.url}
                  onChange={() => setSelectedCharm(charm.url)}
                  disabled={!hasActionData}
                />;

hasActionData ? input : <Tooltip message={Label.NO_ACTIONS} position="left" >{input}</Tooltip>

Reply:
Done! Thanks for showing both ways I can improve this part of the code.